In [44]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')
pd.options.display.float_format = '{:.2f}'.format
df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [45]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [46]:
# Use the describe function to gather some basic statistics
df.describe()

client_id   order_id  order_week  order_year  unit_price  unit_cost  \
count   54639.00   54639.00    54639.00    54639.00    54639.00   54639.00   
mean    54837.87 5470190.03       11.36     2022.99      136.27      99.45   
std     25487.44 2599806.93        7.02        0.08      183.87     133.16   
min     10033.00 1000886.00        1.00     2022.00        0.01       0.01   
25%     33593.00 3196372.00        6.00     2023.00       20.80      14.84   
50%     53305.00 5496966.00       11.00     2023.00       68.31      49.89   
75%     78498.00 7733869.00       17.00     2023.00      173.16     125.57   
max     99984.00 9998480.00       52.00     2023.00     1396.23     846.27   

       unit_weight        qty  line_number  
count     54639.00   54639.00     54639.00  
mean          5.00     570.26         2.98  
std           5.33   18795.52         2.44  
min           0.00       0.00         0.00  
25%           1.45      32.00         1.00  
50%           3.24      68.00         3.00  
75%           6.89     170.00         5.00  
max          46.43 3958244.00         9.00

In [47]:
# Use this space to do any additional research
# and familiarize yourself with the data.
#* -------------------------------------------
# df.info()
# df.head()
# df.sort_values(by='order_week', ascending=False).head()
# df.sort_values(by=['category']).head()
# count_jocabs = df[df[['first''last'] == 'Jacob','Mendoza'].value_counts()
# count_jocabs

"""
  if a specific client needed to be found in this data using their 
  fist and last names a new full_name column will need to be create where
  the first and last (name) columns will be joined
"""
df['full_name'] = df['first']+ ' ' + df['last']
df[df['full_name'] == 'Jacob Mendoza']['full_name'].value_counts()

"""
  in cases where you wish to display a single data point without an index it may 
  be appropriate to use .to_string(index=False). This will convert the data point
  to a string which can then be used seamlessly in a print statement
"""
df['full_name'].head(1).to_string(index=False)

'Donald Harding'

In [48]:
# What three item categories had the most entries?

"""
  in this case we need to count how many times each
  category is mentioned and return the top three most
  mentioned categories. a good method for handling this 
  is .value_count()

  @params subset, normalize, sort, ascending, dropna
  @returns the total times that a given value is
  counted in the data
  .value_count()
"""
categories = df[['category']].value_counts().head(3)
categories

category   
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [49]:
# For the category with the most entries, which subcategory had the most entries?

"""
  now that the most mentioned categories have been identified we need to know which
  of their subcategories is mentioned the most. one way to do this is to use .group_by().
  with this we can aggregate the data using the subcategories as our collections and
  then we just need to count their values.

  @params by, axis, level, as_index, sort, group_keys, observed, dropna
  @returns aggregated data collections
  .groupby()
"""
sub_cats = df[['category','subcategory']].groupby(['subcategory']).value_counts()
sub_cats.head(3)

subcategory        category   
accounting         software       3756
audio visual       electronics     732
bathroom supplies  consumables    6424
Name: count, dtype: int64

In [50]:
# Which five clients had the most entries in the data?
# Store the client ids of those top 5 clients in a list.
top_five = df[['full_name','client_id']].value_counts().head()
top_five

full_name        client_id
Jessica Reyes    33615        220
Angela Everett   66037        211
Bryan Myers      46820        209
Kendra Garrett   24741        207
Alexandra Young  38378        207
Name: count, dtype: int64

In [51]:
# How many total units (the qty column) did the client with the most entries order order?

"""
  There may be times that you only need a single value from a dataframe.
  in some cases it may be best to use .index to allow for finding the value
  using its placement index

  .index[0][0]
"""

top_client = df[['full_name', 'client_id']].value_counts().head(1)
top_client = pd.DataFrame(top_client)
name = top_client.index[0][0]
id = top_client.index[0][1]
total_units = df.loc[df['client_id'] == id, 'qty'].sum()
print(f'The client with the most entries is {name} with a total of {total_units} units bought! Good job {name}! Keep it up!')

The client with the most entries is Jessica Reyes with a total of 64313 units bought! Good job Jessica Reyes! Keep it up!


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [52]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty

"""
  There may be times when it is best to create a new column and store
  the aggregated data of one of more columns
"""
df['total_units_price'] = df['unit_price'] * df['qty']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number           full_name  \
0     1096.80     762.71         7.50  105            1      Donald Harding   
1       24.95      15.09         1.49   21            0       Tiffany Myers   
2       13.52       7.86         1.68   39            6      Shannon Watson   
3       36.42      24.85         1.23   29            3        Nathan Baker   
4      195.10     108.17        46.43   20            1  Christina Schwartz   

   total_units_price  
0          115164.00  
1             523.95  
2             527.28  
3            1056.18  
4            3902.00

In [53]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

"""
  We need to know how much it is going to cost to ship our units and we need to store that
  shipping price. Using an if-else statement should make it possible to add the correct value to the new shipping_price column
"""
if df['unit_weight'].all() >= 50:
  df['shipping_price'] = df['unit_weight'] * 10
else:
  df['shipping_price'] = df['unit_weight'] * 7


In [54]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

"""
  In cases where you are working with floats it may be necessary to reformat pandas.
  To do this you can use pd.options.display.float_format = '{:.2f}'.format
"""
df['total_price'] = df['unit_price'] + df['shipping_price'] * .0925


In [55]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

df['total_processing_cost'] = df.unit_cost + df.qty + df.shipping_price

In [56]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df.unit_price - df.unit_cost


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [57]:
# Check your work using the totals above
"""
  For this we will use a lambda with an axis of 1
"""
df.apply(lambda row: print(f'Order ID {row.order_id} had a total price of ${float(row.total_units_price):,.2f}'), axis=1).dropna()

Order ID 8953482 had a total price of $115,164.00
Order ID 8069089 had a total price of $523.95
Order ID 1902144 had a total price of $527.28
Order ID 9031802 had a total price of $1,056.18
Order ID 1322274 had a total price of $3,902.00
Order ID 1327730 had a total price of $13,481.52
Order ID 8080280 had a total price of $13,751.36
Order ID 8946782 had a total price of $461.78
Order ID 9394118 had a total price of $4,109.14
Order ID 9665852 had a total price of $5,216.64
Order ID 5775924 had a total price of $50,960.00
Order ID 2859900 had a total price of $890.42
Order ID 6059361 had a total price of $130,492.32
Order ID 3470698 had a total price of $1,483.68
Order ID 5953045 had a total price of $6,162.25
Order ID 5014569 had a total price of $869.71
Order ID 2101230 had a total price of $157.78
Order ID 4397711 had a total price of $1,637.09
Order ID 6832072 had a total price of $399.33
Order ID 7573941 had a total price of $46,378.65
Order ID 7490302 had a total price of $13,386.

Series([], dtype: object)

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [65]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.



client_id  total_units
0      58515     31158688
1      37609     31158688
2      57113     31158688
3      46554     31158688
4      92089     31158688

In [59]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 


In [60]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




In [61]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
